### PR TITLE
Fixed #28872 -- Added JSONIn lookup

### PR DESCRIPTION
**Pull request by VCAMP**
https://code.djangoproject.com/ticket/28872

We noticed this bug was still present in `master` and handled it in a way that should be supported in Django 1.11+/Postgres 9.3+. Happy to receive feedback on other ways to solve it/things we might have overlooked in the implementation.

Thanks to Jerome Leclanche for reporting and suggesting a solution.

### CODE REVIEW

File: tests/postgres_tests/test_json.py [hunk 1]
```diff
@@ -299,6 +299,24 @@ def test_regex(self):
     def test_iregex(self):
         self.assertTrue(JSONModel.objects.filter(field__foo__iregex=r'^bAr$').exists())
 
+    def test_deep_lookup_qset_rhs(self):
+        # See ticket #28872
```
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: Ticket references should be reserved for obscure issues (not needed here).

File: tests/postgres_tests/test_json.py [hunk 2]
```diff
@@ -299,6 +299,24 @@ def test_regex(self):
...
+        Character.objects.create(name='bar')
+
```
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: The blank lines aren't needed.

File: tests/postgres_tests/test_json.py [hunk 3]
```diff
@@ -299,6 +299,24 @@ def test_regex(self):
...
+        query = JSONModel.objects.filter(
+            field__foo__in=Character.objects.values_list('name')
+        )
+
+        self.assertSequenceEqual(query, [self.objs[11]])
+
+    def test_deep_lookup_direct_value_rhs(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: This test seems to be passing without the patch (fine to add additional test coverage but we'd rather do it in a separate commit so it's clear which case the patch fixes).

File: django/contrib/postgres/lookups.py [hunk 4]
```diff
@@ -74,3 +74,15 @@ def process_rhs(self, compiler, connection):
...
         # Treat None lookup values as null.
         return ("'null'", []) if result == ('%s', [None]) else result
+
+
+class JSONIn(In):
+    def process_rhs(self, compiler, connection):
+        rhs, rhs_params = super().process_rhs(compiler, connection)
+        if not self.rhs_is_direct_value():
+            # We're doing an explicit type cast here as
+            # Postgres 9.3 doesn't have access to a
```
Comment (timgraham) [inline django/contrib/postgres/lookups.py]: The master branch only supports PostgreSQL 9.4+.

Comment (timgraham) [reply django/contrib/postgres/lookups.py]: Is it worth making the code conditional on a `connection.features` flag so that PostgreSQL 9.5+ can use  `to_jsonb`?

File: tests/postgres_tests/test_json.py (hunk 3 above)
Comment (VCAMP) [reply tests/postgres_tests/test_json.py]: Makes sense - will remove this test from the commit.

File: django/contrib/postgres/lookups.py (hunk 4 above)
Comment (VCAMP) [reply django/contrib/postgres/lookups.py]: According to [someone on the Internet](https://stackoverflow.com/questions/31969901/postgres-json-to-jsonb) there is a non-trivial performance cost when converting from `json` to `jsonb`, so we should avoid doing it unless it's necessary. Happy to make the change if you think the performance cost is going to make a difference here.

Comment (VCAMP) [reply django/contrib/postgres/lookups.py]: Have added a check for Postgres 9.4. It's not explicitly tested atm but I assume in this case it's more of a job for the build system. Let me know if that looks acceptable.

**Comment by timgraham** [comment]
It fails with "more than one row returned by a subquery used as an expression" if `Character.objects.values_list('name')` returns more than 1 row.

**Comment by VCAMP** [comment]
> It fails with "more than one row returned by a subquery used as an expression" if `Character.objects.values_list('name')` returns more than 1 row.

@timgraham you're right, good spot.

Considering that:

1) We can't wrap the whole rhs subquery with a `to_jsonb` call because of the error you just found;
2) We can only add a type cast/call to `to_jsonb` around the column name(s) mentioned by rhs `SELECT` statement (e.g. if the rhs expression is `SELECT U0."name" FROM table_character"` it needs to become `SELECT to_jsonb(U0."name") FROM table_character`) for this to work
3) The rhs subquery can only be one that returns values from a single column (else you get a `TypeError: Cannot use multi-field values as a filter value.`) but it can contain expressions which modify it 

I'm wondering if there's a way to access the rhs subquery from the `IN` lookup before it's compiled into SQL in order to modify it safely? Else we need to do some string munging in the lookup's `process_rhs` method, which doesn't feel especially clean.

**Comment by timgraham** [comment]
I don't have much expertise to provide guidance.

**Comment by VCAMP** [comment]
@timgraham we dug a bit deeper in the code and found a possible workaround. It's not the cleanest solution but considered we're handling a very specific edge case we thought it would not make sense to extend `SQLCompiler.as_sql`. Can you/someone else provide feedback?

**Comment by felixxm** [comment]
This patch need many changes after #12392, also the current master supports only PostgreSQL 9.6+, and as far as I'm concerned this issue is PostgreSQL-specific.
